File: apps-microservices/crawler-service-python/tests/test_domain_fr.py
import sys
import os
import asyncio
import logging

# Ensure 'src' is in python path
current_dir = os.path.dirname(os.path.abspath(__file__))
src_path = os.path.join(current_dir, '..', 'src')
sys.path.insert(0, src_path)

from domain_fr import DomainFR

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(message)s')
logger = logging.getLogger("Test")

async def run_tests():
    logger.info("=== STARTING PYTHON DOMAIN_FR TESTS ===")
    
    # --- PART 1: CONTENT DETECTION ---
    logger.info("\n--- Test 1: Content Language Detection ---")
    domain_fr = DomainFR("http://example.com")
    
    test_cases_content = [
        {
            "name": "HTML Lang Attribute (fr-FR)",
            "content": '<!DOCTYPE html><html lang="fr-FR"><head></head><body></body></html>',
            "expected": "fr"
        },
        {
            "name": "HTML Lang Attribute (fr)",
            "content": '<html lang="fr">',
            "expected": "fr"
        },
        {
            "name": "OpenGraph Locale",
            "content": '<meta property="og:locale" content="fr_FR" />',
            "expected": "fr"
        },
        {
            "name": "Meta Language",
            "content": '<meta name="LANGUAGE" content="fr" />',
            "expected": "fr"
        },
        {
            "name": "HTTP-Equiv",
            "content": '<meta http-equiv="content-language" content="fr" />',
            "expected": "fr"
        },
        {
            "name": "English Content",
            "content": '<html lang="en-US">',
            "expected": "en"
        },
        {
            "name": "No Language Info",
            "content": '<html><body>Hello</body></html>',
            "expected": False
        }
    ]

    for test in test_cases_content:
        result = domain_fr.detect_language(test["content"])
        
        val = result["value"] if result else False
        expected = test["expected"]
        
        is_success = val == expected
        icon = "✅" if is_success else "❌"
        logger.info(f"{icon} [{test['name']}] Expected: {expected} | Got: {val}")

    # --- PART 2: URL PATTERN DETECTION ---
    logger.info("\n--- Test 2: URL Pattern Detection ---")
    
    test_cases_url = [
        {"url": "https://www.example.fr", "name": "TLD .fr", "expected": True},
        {"url": "https://fr.example.com", "name": "Subdomain fr.", "expected": True},
        {"url": "https://www.example.com/fr/page", "name": "Path /fr/", "expected": True},
        {"url": "https://www.example.com/page?lang=fr", "name": "Query lang=fr", "expected": True},
        {"url": "https://www.example.com/page?locale=fr-CA", "name": "Query locale=fr-CA", "expected": True},
        {"url": "https://www.example.com", "name": "Generic .com", "expected": False},
        {"url": "https://www.example.com/en/page", "name": "English Path", "expected": False}
    ]

    for test in test_cases_url:
        # Check URL without tracking redirects (unit test logic)
        result = await DomainFR.check_url(test["url"], track_redirect=False)
        is_ok = result.get("ok", False)
        
        is_success = is_ok == test["expected"]
        icon = "✅" if is_success else "❌"
        logger.info(f"{icon} [{test['name']}] ({test['url']}) Expected: {test['expected']} | Got: {is_ok}")

    logger.info("\n=== TESTS FINISHED ===")

if __name__ == "__main__":
    try:
        asyncio.run(run_tests())
    except KeyboardInterrupt:
        pass

File: apps-microservices/crawler-service/crawler/src/tests/test_domain_fr.ts
import { DetectionLangueClient } from "../class/DetectionLangueClient.js";

const log = (msg: string, success: boolean) => {
    if (success) console.log(`✅ ${msg}`);
    else console.error(`❌ ${msg}`);
};

async function runTests() {
    console.log("=== STARTING DETECTION LANGUE CLIENT TESTS ===");

    // --- PART 1: extractPrimaryMethod (Unit - no API needed) ---
    console.log("\n--- Test 1: extractPrimaryMethod ---");

    const methodCases = [
        // HTML method present → prefer it over URL method
        { input: "langHtml", expected: "langHtml" },
        { input: "langHtml+nlp_confirmed", expected: "langHtml" },
        { input: "direct_match+langHtml+nlp_confirmed", expected: "langHtml" },
        { input: "direct_match+matchMeta+nlp_soft_confirmed", expected: "matchMeta" },
        { input: "pattern_match_path+matchHttpEquiv", expected: "matchHttpEquiv" },
        // No HTML method → fall back to first component
        { input: "matchMeta", expected: "matchMeta" },
        { input: "direct_match+nlp_confirmed", expected: "direct_match" },
        { input: "pattern_match_path+nlp_soft_confirmed", expected: "pattern_match_path" },
        { input: "nlp_confirmed", expected: "nlp_confirmed" },
        { input: "direct_match", expected: "direct_match" },
        // Edge case: empty/undefined method guard
        { input: "", expected: "" },
    ];

    for (const test of methodCases) {
        const result = DetectionLangueClient.extractPrimaryMethod(test.input);
        log(
            `extractPrimaryMethod("${test.input}") Expected: "${test.expected}" | Got: "${result}"`,
            result === test.expected
        );
    }

    // --- PART 1b: requiresNlpValidation (Unit - no API needed) ---
    console.log("\n--- Test 1b: requiresNlpValidation ---");

    const nlpValidationCases = [
        // HTML-based methods (whitelist) → forced_method works, no NLP needed
        { input: "langHtml", expected: false },
        { input: "matchMeta", expected: false },
        { input: "matchHttpEquiv", expected: false },
        // URL-based methods → forced_method useless, needs NLP
        { input: "direct_match", expected: true },
        { input: "pattern_match_path", expected: true },
        { input: "pattern_match_query", expected: true },
        // NLP-only methods → needs NLP
        { input: "nlp_confirmed", expected: true },
        { input: "nlp_soft_confirmed", expected: true },
        { input: "french_lexical_signal", expected: true },
        { input: "alternative_link_validated", expected: true },
        // Edge cases: nlp_override_* and unknown methods also need NLP (whitelist approach)
        { input: "nlp_override_langHtml", expected: true },
        { input: "no_redirect", expected: true },
        { input: "some_future_method", expected: true },
        { input: "", expected: true },
    ];

    for (const test of nlpValidationCases) {
        const result = DetectionLangueClient.requiresNlpValidation(test.input);
        log(
            `requiresNlpValidation("${test.input}") Expected: ${test.expected} | Got: ${result}`,
            result === test.expected
        );
    }

    // --- PART 2: API Integration Tests (require running API) ---
    const apiUrl = process.env.DETECTION_LANGUE_API_URL || "http://api-detection-langue-fr-service:8999";
    const client = new DetectionLangueClient(apiUrl);

    // Check if the API is reachable before running integration tests
    try {
        const healthCheck = await client.checkUrl("https://www.example.fr", false);
        console.log("\n--- Test 2: check-url endpoint ---");

        const testCasesUrl = [
            { url: "https://www.example.fr", name: "TLD .fr", expected: true },
            { url: "https://fr.example.com", name: "Subdomain fr.", expected: true },
            { url: "https://www.example.com/fr/page", name: "Path /fr/", expected: true },
            { url: "https://www.example.com/page?lang=fr", name: "Query lang=fr", expected: true },
            { url: "https://www.example.com", name: "Generic .com", expected: false },
            { url: "https://www.example.com/en/page", name: "English Path", expected: false },
        ];

        for (const test of testCasesUrl) {
            const result = await client.checkUrl(test.url, false);
            log(
                `[${test.name}] (${test.url}) Expected: ${test.expected} | Got: ${result.ok}`,
                result.ok === test.expected
            );
        }

        // --- Test 3: detect endpoint with HTML content ---
        console.log("\n--- Test 3: detect endpoint (with HTML content) ---");

        const testCasesDetect = [
            {
                name: "French HTML lang",
                url: "https://www.example.com",
                content: '<!DOCTYPE html><html lang="fr-FR"><head></head><body><p>Bonjour le monde</p></body></html>',
                expected: true,
            },
            {
                name: "English HTML lang",
                url: "https://www.example.com",
                content: '<!DOCTYPE html><html lang="en-US"><head></head><body><p>Hello world</p></body></html>',
                expected: false,
            },
        ];

        for (const test of testCasesDetect) {
            const result = await client.detect(test.url, test.content, {
                mode: "simple",
                useNlpDetection: false,
            });
            log(
                `[${test.name}] Expected ok=${test.expected} | Got ok=${result.ok} (method: ${result.method})`,
                result.ok === test.expected
            );
        }
    } catch (e: any) {
        console.warn(`\n⚠️  API not reachable at ${apiUrl} — skipping integration tests.`);
        console.warn(`   Error: ${e.message}`);
        console.warn(`   Start the API with: docker compose --profile app up api-detection-langue-fr-service`);
    }

    console.log("\n=== TESTS FINISHED ===");
}

runTests().catch(console.error);
